registry/internal/tools/finder/github.go

```diff
@@ -31,26 +32,32 @@ type github struct {
 }
 
 func (g *github) Request(source string) (*Result, error) {
-	tagsURI, err := g.tagsURI(source)
+	namespace, project, err := g.parseSource(source)
 	if err != nil {
 		return nil, err
 	}
 
+	tagsURI := g.tagsURI(namespace, project)
+
 	g.log.Tracef("requesting tags from URI: %s", tagsURI)
 
 	tags, err := g.requestTags(tagsURI)
 	if err != nil {
 		return nil, err
 	}
 
-	headURI, err := g.headURI(source)
+	gomodURI := g.gomodURI(namespace, project)
+	g.log.Tracef("requesting go.mod from URI: %s", gomodURI)
+	hasGomod, err := g.hasGomod(gomodURI)
```

Comment:
I'm not sure about the validity of this approach for determining if a project is go-mod compliant or not.

Reply:
Unfortunately the story gets more complex; I created 
https://github.com/shoenig/testnest 
as a little demo of where this doesn't work (i.e. any module which puts its `go.mod` file other than in the root of the project directory)

The go proxy is fine with this module: 
```
$ curl "proxy.golang.org/github.com/shoenig/testnest/@v/list"
v0.0.0
```
But the equivelent request made to the URI create with `gomodURI` isn't going to see it:
```
$ curl -w "\n%{response_code}\n" "https://github.com/repos/shoenig/testnest/contents/go.mod"
Not Found
404
```

Reply:
Getting even more complex, the Go compiler supports vanity URLs which are like 99.9% of the time backed by github repositories. One of the more common use cases for this feature would be to help with the kubernetes modules, which have various vanity URLs using this technique. 

E.g.
The module:
```
$ curl "proxy.golang.org/gophers.dev/pkgs/ignore/@v/list"
v0.0.0
v0.1.0
v0.2.0
```
In reality, that URL is nothing more than a `go-get` redirect (https://golang.org/cmd/go/#hdr-Remote_import_paths)
```
[p1 ~] $ curl -sL  gophers.dev/pkgs/ignore
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta
            name="go-import"
            content="gophers.dev/pkgs/ignore git https://github.com/shoenig/ignore"
    >
    <title>go-get</title>
</head>
</html>
```

Reply:
While there is code in modprox already for resolving redirects and for DLing archives (which we would need to do to properly detect the presence of a `go.mod` file), it might be worth investigating if we can leverage the proxy.golang.org API instead, which would hide all that gross plumbing.

Reply:
I've looked at the proxy.golang.org api and I think I have a reasonable approach.

We only care about module-incompatibility in the case where the repo is module-incompatible _and_ has a major semver tag >= 2 (e.g github.com/sourcegraph/go-langserver), so that we add the `+incompatible` suffix to both the "latest commit" pseudo-version and to any semver >= 2.

So if we get the list of versions from the API and see any versions with `+incompatible`, that's all we need to know. For module-incompatible repos that don't have a semver >= 2, we don't care ; we can treat them as compatible.